README.md
Swauth
------

An Auth Service for Swift as WSGI Middleware that uses Swift itself as a
backing store. Docs at: <https://swauth.readthedocs.io/> or ask in #openstack-swauth on
freenode [IRC](http://eavesdrop.openstack.org/irclogs/%23openstack-swauth/).

See also <https://github.com/openstack/keystone> for the standard OpenStack
auth service.


NOTE
----

**Be sure to review the docs at:
<https://swauth.readthedocs.io/>**


Quick Install
-------------

1) Install Swauth with ``sudo python setup.py install`` or ``sudo python
   setup.py develop`` or via whatever packaging system you may be using.

2) Alter your proxy-server.conf pipeline to have swauth instead of tempauth:

    Was:

        [pipeline:main]
        pipeline = catch_errors cache tempauth proxy-server

    Change To:

        [pipeline:main]
        pipeline = catch_errors cache swauth proxy-server

3) Add to your proxy-server.conf the section for the Swauth WSGI filter:

    [filter:swauth]
    use = egg:swauth#swauth
    set log_name = swauth
    super_admin_key = swauthkey

4) Be sure your proxy server allows account management:

    [app:proxy-server]
    ...
    allow_account_management = true

5) Restart your proxy server ``swift-init proxy reload``

6) Initialize the Swauth backing store in Swift ``swauth-prep -K swauthkey``

7) Add an account/user ``swauth-add-user -A http://127.0.0.1:8080/auth/ -K
   swauthkey -a test tester testing``

8) Ensure it works ``swift -A http://127.0.0.1:8080/auth/v1.0 -U test:tester -K
   testing stat -v``


Web Admin Install
-----------------

1)  If you installed from packages, you'll need to cd to the webadmin directory
    the package installed. This is ``/usr/share/doc/python-swauth/webadmin``
    with the Lucid packages. If you installed from source, you'll need to cd to
    the webadmin directory in the source directory.

2)  Upload the Web Admin files with ``swift -A http://127.0.0.1:8080/auth/v1.0
    -U .super_admin:.super_admin -K swauthkey upload .webadmin .``

3)  Open ``http://127.0.0.1:8080/auth/`` in your browser.


Swift3 Middleware Compatibility
-------------------------------
[**Swift3 middleware**](https://github.com/openstack/swift3) can be used with
swauth when `auth_type` in swauth is configured to be *Plaintext* (default).

    [pipeline:main]
    pipeline = catch_errors cache swift3 swauth proxy-server

It can be used with `auth_type` set to Sha1/Sha512 too but with certain caveats.
Refer to swift3 compatibility [section](https://swauth.readthedocs.io/en/latest/#swift3-middleware-compatibility)
in documentation for further details


CHANGELOG
swauth (1.3.0)

    [SECURITY] Stop using client headers for cross-middleware communication
    WARNING: You need to upgrade Swift3 to at least 1.12

    [SECURITY] Hash token before storing it in Swift (CVE-2017-16613)
    WARNING: In deployments without memcached this patch logs out all users
    because tokens became invalid.

swauth (1.2.0)

    Allow to set password by hash

    Allow to set hash salt in config for S3 compatibility

    Due to security reason, S3 support is disabled by default

    Salt is not included in S3 HMAC computation

    Use correct content type on JSON responses

    Fix changing of auth_type in existing deployments

    Remove outdated locale

swauth (1.1.0)

    This is first release after move to OpenStack Infra

    Allow users to change their own password/key

    Add support for storage policy

    Show password prompt if key is not specified

    Allow to use Keystone at same time

    Support SHA512 for password hashing

    Code cleanup

    Bugfixies a security fixies

swauth (1.0.8)

   Added request.environ[reseller_request] = True if request is coming from an
   user in .reseller_admin group

   Fixed to work with newer Swift versions whose memcache clients require a
   time keyword argument when the older versions required a timeout keyword
   argument.

swauth (1.0.7)

    New X-Auth-Token-Lifetime header a user can set to how long they'd like
    their token to be good for.

    New max_token_life config value for capping the above.

    New X-Auth-Token-Expires header returned with the get token request.

    Switchover to swift.common.swob instead of WebOb; requires Swift >= 1.7.6
    now.

swauth (1.0.6)

    Apparently I haven't been keeping up with this CHANGELOG. I'll try to be
    better onward.

    This release added passing OPTIONS requests through untouched, needed for
    CORS support in Swift.

    Also, Swauth is a bit more restrictive in deciding when it's the definitive
    auth for a request.

swauth (1.0.3-dev)

    This release is still under development. A full change log will be made at
    release. Until then, you can see what has changed with:

    git log 1.0.2..HEAD

swauth (1.0.2)

    Fixed bug rejecting requests when using multiple instances of Swauth or
    Swauth with other auth services.

    Fixed bug interpreting URL-encoded user names and keys.

    Added support for the Swift container sync feature.

    Allowed /not/ setting super_admin_key to disable Swauth administration
    features.

    Added swauth_remote mode so the Swauth middleware for one Swift cluster
    could be pointing to the Swauth service on another Swift cluster, sharing
    account/user data sets.

    Added ability to purge stored tokens.

    Added API documentation for internal Swauth API.

swauth (1.0.1)

    Initial release after separation from Swift.


test/__init__.py
# See http://code.google.com/p/python-nose/issues/detail?id=373
# The code below enables nosetests to work with i18n _() blocks

import six.moves.builtins as __builtin__

setattr(__builtin__, '_', lambda x: x)


test/unit/test_authtypes.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Pablo Llopis 2011

import mock
from swauth import authtypes
import unittest


class TestValidation(unittest.TestCase):
    def test_validate_creds(self):
        creds = 'plaintext:keystring'
        creds_dict = dict(type='plaintext', salt=None, hash='keystring')
        auth_encoder, parsed_creds = authtypes.validate_creds(creds)
        self.assertEqual(parsed_creds, creds_dict)
        self.assertTrue(isinstance(auth_encoder, authtypes.Plaintext))

        creds = 'sha1:salt$d50dc700c296e23ce5b41f7431a0e01f69010f06'
        creds_dict = dict(type='sha1', salt='salt',
                          hash='d50dc700c296e23ce5b41f7431a0e01f69010f06')
        auth_encoder, parsed_creds = authtypes.validate_creds(creds)
        self.assertEqual(parsed_creds, creds_dict)
        self.assertTrue(isinstance(auth_encoder, authtypes.Sha1))

        creds = ('sha512:salt$482e73705fac6909e2d78e8bbaf65ac3ca1473'
                 '8f445cc2367b7daa3f0e8f3dcfe798e426b9e332776c8da59c'
                 '0c11d4832931d1bf48830f670ecc6ceb04fbad0f')
        creds_dict = dict(type='sha512', salt='salt',
                          hash='482e73705fac6909e2d78e8bbaf65ac3ca1473'
                               '8f445cc2367b7daa3f0e8f3dcfe798e426b9e3'
                               '32776c8da59c0c11d4832931d1bf48830f670e'
                               'cc6ceb04fbad0f')
        auth_encoder, parsed_creds = authtypes.validate_creds(creds)
        self.assertEqual(parsed_creds, creds_dict)
        self.assertTrue(isinstance(auth_encoder, authtypes.Sha512))

    def test_validate_creds_fail(self):
        # wrong format, missing `:`
        creds = 'unknown;keystring'
        self.assertRaisesRegex(ValueError, "Missing ':' in .*",
                               authtypes.validate_creds, creds)
        # unknown auth_type
        creds = 'unknown:keystring'
        self.assertRaisesRegex(ValueError, "Invalid auth_type: .*",
                               authtypes.validate_creds, creds)
        # wrong plaintext keystring
        creds = 'plaintext:'
        self.assertRaisesRegex(ValueError, "Key must have non-zero length!",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, missing `$`
        creds = 'sha1:saltkeystring'
        self.assertRaisesRegex(ValueError, "Missing '\$' in .*",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, missing salt
        creds = 'sha1:$hash'
        self.assertRaisesRegex(ValueError, "Salt must have non-zero length!",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, missing hash
        creds = 'sha1:salt$'
        self.assertRaisesRegex(ValueError, "Hash must have 40 chars!",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, short hash
        creds = 'sha1:salt$short_hash'
        self.assertRaisesRegex(ValueError, "Hash must have 40 chars!",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, wrong format
        creds = 'sha1:salt$' + "z" * 40
        self.assertRaisesRegex(ValueError, "Hash must be hexadecimal!",
                               authtypes.validate_creds, creds)
        # wrong sha512 format, missing `$`
        creds = 'sha512:saltkeystring'
        self.assertRaisesRegex(ValueError, "Missing '\$' in .*",
                               authtypes.validate_creds, creds)
        # wrong sha512 format, missing salt
        creds = 'sha512:$hash'
        self.assertRaisesRegex(ValueError, "Salt must have non-zero length!",
                               authtypes.validate_creds, creds)
        # wrong sha512 format, missing hash
        creds = 'sha512:salt$'
        self.assertRaisesRegex(ValueError, "Hash must have 128 chars!",
                               authtypes.validate_creds, creds)
        # wrong sha512 format, short hash
        creds = 'sha512:salt$short_hash'
        self.assertRaisesRegex(ValueError, "Hash must have 128 chars!",
                               authtypes.validate_creds, creds)
        # wrong sha1 format, wrong format
        creds = 'sha512:salt$' + "z" * 128
        self.assertRaisesRegex(ValueError, "Hash must be hexadecimal!",
                               authtypes.validate_creds, creds)


class TestPlaintext(unittest.TestCase):

    def setUp(self):
        self.auth_encoder = authtypes.Plaintext()

    def test_plaintext_encode(self):
        enc_key = self.auth_encoder.encode('keystring')
        self.assertEqual('plaintext:keystring', enc_key)

    def test_plaintext_valid_match(self):
        creds = 'plaintext:keystring'
        match = self.auth_encoder.match('keystring', creds)
        self.assertEqual(match, True)

    def test_plaintext_invalid_match(self):
        creds = 'plaintext:other-keystring'
        match = self.auth_encoder.match('keystring', creds)
        self.assertEqual(match, False)


class TestSha1(unittest.TestCase):

    def setUp(self):
        self.auth_encoder = authtypes.Sha1()
        self.auth_encoder.salt = 'salt'

    @mock.patch('swauth.authtypes.os')
    def test_sha1_encode(self, os):
        os.urandom.return_value.encode.return_value.rstrip \
            .return_value = 'salt'
        enc_key = self.auth_encoder.encode('keystring')
        self.assertEqual('sha1:salt$d50dc700c296e23ce5b41f7431a0e01f69010f06',
                         enc_key)

    def test_sha1_valid_match(self):
        creds = 'sha1:salt$d50dc700c296e23ce5b41f7431a0e01f69010f06'
        creds_dict = dict(type='sha1', salt='salt',
                          hash='d50dc700c296e23ce5b41f7431a0e01f69010f06')
        match = self.auth_encoder.match('keystring', creds, **creds_dict)
        self.assertEqual(match, True)

    def test_sha1_invalid_match(self):
        creds = 'sha1:salt$deadbabedeadbabedeadbabec0ffeebadc0ffeee'
        creds_dict = dict(type='sha1', salt='salt',
                          hash='deadbabedeadbabedeadbabec0ffeebadc0ffeee')
        match = self.auth_encoder.match('keystring', creds, **creds_dict)
        self.assertEqual(match, False)

        creds = 'sha1:salt$d50dc700c296e23ce5b41f7431a0e01f69010f06'
        creds_dict = dict(type='sha1', salt='salt',
                          hash='d50dc700c296e23ce5b41f7431a0e01f69010f06')
        match = self.auth_encoder.match('keystring2', creds, **creds_dict)
        self.assertEqual(match, False)


class TestSha512(unittest.TestCase):

    def setUp(self):
        self.auth_encoder = authtypes.Sha512()
        self.auth_encoder.salt = 'salt'

    @mock.patch('swauth.authtypes.os')
    def test_sha512_encode(self, os):
        os.urandom.return_value.encode.return_value.rstrip \
            .return_value = 'salt'
        enc_key = self.auth_encoder.encode('keystring')
        self.assertEqual('sha512:salt$482e73705fac6909e2d78e8bbaf65ac3ca1473'
                         '8f445cc2367b7daa3f0e8f3dcfe798e426b9e332776c8da59c'
                         '0c11d4832931d1bf48830f670ecc6ceb04fbad0f', enc_key)

    def test_sha512_valid_match(self):
        creds = ('sha512:salt$482e73705fac6909e2d78e8bbaf65ac3ca14738f445cc2'
                 '367b7daa3f0e8f3dcfe798e426b9e332776c8da59c0c11d4832931d1bf'
                 '48830f670ecc6ceb04fbad0f')
        creds_dict = dict(type='sha512', salt='salt',
                          hash='482e73705fac6909e2d78e8bbaf65ac3ca14738f445cc2'
                               '367b7daa3f0e8f3dcfe798e426b9e332776c8da59c0c11'
                               'd4832931d1bf48830f670ecc6ceb04fbad0f')
        match = self.auth_encoder.match('keystring', creds, **creds_dict)
        self.assertEqual(match, True)

    def test_sha512_invalid_match(self):
        creds = ('sha512:salt$deadbabedeadbabedeadbabedeadbabedeadbabedeadba'
                 'bedeadbabedeadbabedeadbabedeadbabedeadbabedeadbabedeadbabe'
                 'c0ffeebadc0ffeeec0ffeeba')
        creds_dict = dict(type='sha512', salt='salt',
                          hash='deadbabedeadbabedeadbabedeadbabedeadbabedeadba'
                               'bedeadbabedeadbabedeadbabedeadbabedeadbabedead'
                               'babedeadbabec0ffeebadc0ffeeec0ffeeba')
        match = self.auth_encoder.match('keystring', creds, **creds_dict)
        self.assertEqual(match, False)

        creds = ('sha512:salt$482e73705fac6909e2d78e8bbaf65ac3ca14738f445cc2'
                 '367b7daa3f0e8f3dcfe798e426b9e332776c8da59c0c11d4832931d1bf'
                 '48830f670ecc6ceb04fbad0f')
        creds_dict = dict(type='sha512', salt='salt',
                          hash='482e73705fac6909e2d78e8bbaf65ac3ca14738f445cc2'
                               '367b7daa3f0e8f3dcfe798e426b9e332776c8da59c0c11'
                               'd4832931d1bf48830f670ecc6ceb04fbad0f')
        match = self.auth_encoder.match('keystring2', creds, **creds_dict)
        self.assertEqual(match, False)

if __name__ == '__main__':
    unittest.main()


swauth/swift_version.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import swift


MAJOR = None
MINOR = None
REVISION = None
FINAL = None


def parse(value):
    parts = value.split('.')
    if parts[-1].endswith('-dev'):
        final = False
        parts[-1] = parts[-1][:-4]
    else:
        final = True
    major = int(parts.pop(0))
    minor = int(parts.pop(0))
    if parts:
        revision = int(parts.pop(0).split('-', 1)[0])
    else:
        revision = 0
    return major, minor, revision, final


def newer_than(value):
    global MAJOR, MINOR, REVISION, FINAL
    try:
        major, minor, revision, final = parse(value)
        if MAJOR is None:
            MAJOR, MINOR, REVISION, FINAL = parse(swift.__version__)
        if MAJOR < major:
            return False
        elif MAJOR == major:
            if MINOR < minor:
                return False
            elif MINOR == minor:
                if REVISION < revision:
                    return False
                elif REVISION == revision:
                    if not FINAL or final:
                        return False
    except Exception:
        # Unable to detect if it's newer, better to fail
        return False
    return True


def at_least(value):
    global MAJOR, MINOR, REVISION, FINAL
    try:
        major, minor, revision, final = parse(value)
        if MAJOR is None:
            MAJOR, MINOR, REVISION, FINAL = parse(swift.__version__)
        if MAJOR < major:
            return False
        elif MAJOR == major:
            if MINOR < minor:
                return False
            elif MINOR == minor:
                if REVISION < revision:
                    return False
                elif REVISION == revision:
                    if not FINAL and final:
                        return False
    except Exception:
        # Unable to detect if it's newer, better to fail
        return False
    return True


swauth/authtypes.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Pablo Llopis 2011


"""This module hosts available auth types for encoding and matching user keys.
For adding a new auth type, simply write a class that satisfies the following
conditions:

- For the class name, capitalize first letter only. This makes sure the user
  can specify an all-lowercase config option such as "plaintext" or "sha1".
  Swauth takes care of capitalizing the first letter before instantiating it.
- Write an encode(key) method that will take a single argument, the user's key,
  and returns the encoded string. For plaintext, this would be
  "plaintext:<key>"
- Write a match(key, creds) method that will take two arguments: the user's
  key, and the user's retrieved credentials. Return a boolean value that
  indicates whether the match is True or False.
"""

import hashlib
import os
import string
import sys


#: Maximum length any valid token should ever be.
MAX_TOKEN_LENGTH = 5000


def validate_creds(creds):
    """Parse and validate user credentials whether format is right

    :param creds: User credentials
    :returns: Auth_type class instance and parsed user credentials in dict
    :raises ValueError: If credential format is wrong (eg: bad auth_type)
    """
    try:
        auth_type, auth_rest = creds.split(':', 1)
    except ValueError:
        raise ValueError("Missing ':' in %s" % creds)
    authtypes = sys.modules[__name__]
    auth_encoder = getattr(authtypes, auth_type.title(), None)
    if auth_encoder is None:
        raise ValueError('Invalid auth_type: %s' % auth_type)
    auth_encoder = auth_encoder()
    parsed_creds = dict(type=auth_type, salt=None, hash=None)
    parsed_creds.update(auth_encoder.validate(auth_rest))
    return auth_encoder, parsed_creds


class Plaintext(object):
    """Provides a particular auth type for encoding format for encoding and
    matching user keys.

    This class must be all lowercase except for the first character, which
    must be capitalized. encode and match methods must be provided and are
    the only ones that will be used by swauth.
    """
    def encode(self, key):
        """Encodes a user key into a particular format. The result of this method
        will be used by swauth for storing user credentials.

        :param key: User's secret key
        :returns: A string representing user credentials
        """
        return "plaintext:%s" % key

    def match(self, key, creds, **kwargs):
        """Checks whether the user-provided key matches the user's credentials

        :param key: User-supplied key
        :param creds: User's stored credentials
        :param kwargs: Extra keyword args for compatibility reason with
                       other auth_type classes
        :returns: True if the supplied key is valid, False otherwise
        """
        return self.encode(key) == creds

    def validate(self, auth_rest):
        """Validate user credentials whether format is right for Plaintext

        :param auth_rest: User credentials' part without auth_type
        :return: Dict with a hash part of user credentials
        :raises ValueError: If credentials' part has zero length
        """
        if len(auth_rest) == 0:
            raise ValueError("Key must have non-zero length!")
        return dict(hash=auth_rest)


class Sha1(object):
    """Provides a particular auth type for encoding format for encoding and
    matching user keys.

    This class must be all lowercase except for the first character, which
    must be capitalized. encode and match methods must be provided and are
    the only ones that will be used by swauth.
    """

    def encode_w_salt(self, salt, key):
        """Encodes a user key with salt into a particular format. The result of
        this method will be used internally.

        :param salt: Salt for hashing
        :param key: User's secret key
        :returns: A string representing user credentials
        """
        enc_key = '%s%s' % (salt, key)
        enc_val = hashlib.sha1(enc_key).hexdigest()
        return "sha1:%s$%s" % (salt, enc_val)

    def encode(self, key):
        """Encodes a user key into a particular format. The result of this method
        will be used by swauth for storing user credentials.

        If salt is not manually set in conf file, a random salt will be
        generated and used.

        :param key: User's secret key
        :returns: A string representing user credentials
        """
        salt = self.salt or os.urandom(32).encode('base64').rstrip()
        return self.encode_w_salt(salt, key)

    def match(self, key, creds, salt, **kwargs):
        """Checks whether the user-provided key matches the user's credentials

        :param key: User-supplied key
        :param creds: User's stored credentials
        :param salt: Salt for hashing
        :param kwargs: Extra keyword args for compatibility reason with
                       other auth_type classes
        :returns: True if the supplied key is valid, False otherwise
        """
        return self.encode_w_salt(salt, key) == creds

    def validate(self, auth_rest):
        """Validate user credentials whether format is right for Sha1

        :param auth_rest: User credentials' part without auth_type
        :return: Dict with a hash and a salt part of user credentials
        :raises ValueError: If credentials' part doesn't contain delimiter
                            between a salt and a hash.
        """
        try:
            auth_salt, auth_hash = auth_rest.split('$')
        except ValueError:
            raise ValueError("Missing '$' in %s" % auth_rest)

        if len(auth_salt) == 0:
            raise ValueError("Salt must have non-zero length!")
        if len(auth_hash) != 40:
            raise ValueError("Hash must have 40 chars!")
        if not all(c in string.hexdigits for c in auth_hash):
            raise ValueError("Hash must be hexadecimal!")

        return dict(salt=auth_salt, hash=auth_hash)


class Sha512(object):
    """Provides a particular auth type for encoding format for encoding and
    matching user keys.

    This class must be all lowercase except for the first character, which
    must be capitalized. encode and match methods must be provided and are
    the only ones that will be used by swauth.
    """

    def encode_w_salt(self, salt, key):
        """Encodes a user key with salt into a particular format. The result of
        this method will be used internal.

        :param salt: Salt for hashing
        :param key: User's secret key
        :returns: A string representing user credentials
        """
        enc_key = '%s%s' % (salt, key)
        enc_val = hashlib.sha512(enc_key).hexdigest()
        return "sha512:%s$%s" % (salt, enc_val)

    def encode(self, key):
        """Encodes a user key into a particular format. The result of this method
        will be used by swauth for storing user credentials.

        If salt is not manually set in conf file, a random salt will be
        generated and used.

        :param key: User's secret key
        :returns: A string representing user credentials
        """
        salt = self.salt or os.urandom(32).encode('base64').rstrip()
        return self.encode_w_salt(salt, key)

    def match(self, key, creds, salt, **kwargs):
        """Checks whether the user-provided key matches the user's credentials

        :param key: User-supplied key
        :param creds: User's stored credentials
        :param salt: Salt for hashing
        :param kwargs: Extra keyword args for compatibility reason with
                       other auth_type classes
        :returns: True if the supplied key is valid, False otherwise
        """
        return self.encode_w_salt(salt, key) == creds

    def validate(self, auth_rest):
        """Validate user credentials whether format is right for Sha512

        :param auth_rest: User credentials' part without auth_type
        :return: Dict with a hash and a salt part of user credentials
        :raises ValueError: If credentials' part doesn't contain delimiter
                            between a salt and a hash.
        """
        try:
            auth_salt, auth_hash = auth_rest.split('$')
        except ValueError:
            raise ValueError("Missing '$' in %s" % auth_rest)

        if len(auth_salt) == 0:
            raise ValueError("Salt must have non-zero length!")
        if len(auth_hash) != 128:
            raise ValueError("Hash must have 128 chars!")
        if not all(c in string.hexdigits for c in auth_hash):
            raise ValueError("Hash must be hexadecimal!")

        return dict(salt=auth_salt, hash=auth_hash)
